denokv/main.rs

```diff
@@ -0,0 +1,584 @@
+use denokv_proto::Mutation;
+use denokv_proto::MutationKind;
+use denokv_proto::ReadRange;
+use denokv_proto::SnapshotReadOptions;
+use denokv_sqlite::Connection;
+use denokv_sqlite::Sqlite;
+use log::error;
+use log::info;
+use prost::DecodeError;
+use rand::SeedableRng;
+use thiserror::Error;
+use time::utc_now;
+use tokio::sync::Notify;
+use uuid::Uuid;
+
+mod config;
+mod limits;
+mod time;
+
+#[derive(Clone)]
+struct AppState {
+  sqlite: Sqlite,
+  access_token: &'static str,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+  let config = Config::parse();
+  std::env::set_var("RUST_LOG", "info");
+  env_logger::init();
+
+  let path = Path::new(&config.sqlite_path);
+  let (sqlite, path) = open_sqlite(path)?;
+  info!("Opened database at {}", path);
+
+  let access_token = Box::leak(config.access_token.into_boxed_str());
+
+  let state = AppState {
+    sqlite,
+    access_token,
+  };
+
+  let v1 = Router::new()
+    .route("/snapshot_read", post(snapshot_read_endpoint))
+    .route("/atomic_write", post(atomic_write_endpoint))
+    .route_layer(middleware::from_fn_with_state(
+      state.clone(),
+      authentication_middleware,
+    ));
+
+  let app = Router::new()
+    .route("/", post(metadata_endpoint))
+    .nest("/v1", v1)
+    .fallback(fallback_handler)
+    .with_state(state);
+
+  let listener = std::net::TcpListener::bind(config.addr)
+    .context("Failed to start server")?;
+  info!("Listening on http://{}", listener.local_addr().unwrap());
+
+  axum::Server::from_tcp(listener)?
+    .serve(app.into_make_service())
+    .await?;
+
+  Ok(())
+}
+
+fn open_sqlite(path: &Path) -> Result<(Sqlite, String), anyhow::Error> {
+  let conn = Connection::open(path)?;
+  let notify = Arc::new(Notify::new());
+  let rng: Box<_> = Box::new(rand::rngs::StdRng::from_entropy());
+  let path = conn.path().unwrap().to_owned();
+  let sqlite = Sqlite::new(conn, notify, rng)?;
+  Ok((sqlite, path))
+}
+
+#[axum::debug_handler]
+async fn metadata_endpoint(
+  State(state): State<AppState>,
+  headers: HeaderMap,
+  maybe_req: Option<Json<MetadataExchangeRequest>>,
+) -> Result<Json<DatabaseMetadata>, ApiError> {
+  let Some(Json(req)) = maybe_req else {
+    return Err(ApiError::MinumumProtocolVersion);
+  };
+  if !req.supported_versions.contains(&2) {
+    return Err(ApiError::NoMatchingProtocolVersion);
+  }
+  let Some(authorization) =
+    headers.get("authorization").and_then(|v| v.to_str().ok())
+  else {
+    return Err(ApiError::MalformedAuthorizationHeader);
+  };
+  let Some((bearer, token)) = authorization.split_once(' ') else {
+    return Err(ApiError::MalformedAuthorizationHeader);
+  };
+  if bearer.to_lowercase() != "bearer" || token != state.access_token {
+    return Err(ApiError::InvalidAccessToken);
+  }
+  let expires_at = utc_now() + Duration::days(1);
+  Ok(Json(DatabaseMetadata {
+    version: 2,
+    database_id: Uuid::nil(),
+    endpoints: vec![EndpointInfo {
```

Comment:
Advertise Alt-Svc?

Reply:
For what? h3?

Reply:
Yes, and turn on `http2_prior_knowledge` by default

Reply:
So you want to add an `altSvc` entry to `EndpointInfo`?

Reply:
Either that or we should support DNS SVCB/HTTPS records

Reply:
This can be another patch though - it's a protocol change